src/watertap_contrib/seto/costing/units/chemical_softening_0D.py

```diff
@@ -0,0 +1,826 @@
+    )
+
+    blk.sludge_disposal_cost = Var(
+        initialize=35,
+        units=costing.base_currency / pyunits.tons,
+        doc="Cost of sludge disposal $/tonne",
+    )
+
+    blk.MgCl2_cost = Var(
+        initialize=1.5,
+        units=costing.base_currency / pyunits.kg,
+        doc="Cost of MgCl2 $/kg",
+    )
+
+    blk.Na2CO3_cost = Var(
+        initialize=0.65,
+        units=costing.base_currency / pyunits.kg,
+        doc="Cost of soda ash $/kg",
+    )
+
+    blk.fix_all_vars()
+
+
+@register_costing_parameter_block(
+    build_rule=build_chem_softening_cost_param_block,
+    parameter_block_name="chemical_softening",
+)
+def cost_chem_softening(blk):
+
+    """
+    Capital and operating costs for excess lime and soda softening
+    """
+    chem_soft = blk.costing_package.chemical_softening
+    make_capital_cost_var(blk)
+    make_fixed_operating_cost_var(blk)
+
+    blk.mixer_energy = Var(
+        initialize=1000,
+        units=pyunits.W,
+        domain=NonNegativeReals,
+        doc="Power consumption for rapid mix",
+    )
+
+    blk.floc_energy = Var(
+        initialize=100,
+        units=pyunits.W,
+        domain=NonNegativeReals,
+        doc="Power consumption for rapid mix",
+    )
+
+    blk.mixer_power = Var(
+        initialize=100,
+        domain=NonNegativeReals,
+        units=pyunits.kWh / pyunits.day,
+        doc="Power consumption for rapid mix",
+    )
+
+    blk.floc_power = Var(
+        initialize=100,
+        domain=NonNegativeReals,
+        units=pyunits.kWh / pyunits.day,
+        doc="Power consumption for rapid mix",
+    )
+
+    blk.electricity_flow = Var(
+        initialize=100,
+        domain=NonNegativeReals,
+        units=pyunits.kWh / pyunits.day,
+        doc="Total electricity consumption per day",
+    )
+
+    # region Capital cost components
+    blk.mix_tank_capital_cost = Var(
+        initialize=1e6,
+        domain=NonNegativeReals,
+        units=blk.costing_package.base_currency,
+        doc="Capital Cost of mixing tank",
+    )
+
+    blk.floc_tank_capital_cost = Var(
+        initialize=1e6,
+        domain=NonNegativeReals,
+        units=blk.costing_package.base_currency,
+        doc="Capital cost of flocculation tank",
+    )
+
+    blk.sed_basin_capital_cost = Var(
+        initialize=1e6,
+        doc="Capital cost of sedimentation basin",
+        domain=NonNegativeReals,
+        units=blk.costing_package.base_currency,
+    )
+
+    blk.recarb_basin_capital_cost = Var(
+        initialize=1e6,
+        doc="Capital cost of recarbonation basin",
+        domain=NonNegativeReals,
+        units=blk.costing_package.base_currency,
+    )
+
+    blk.recarb_basin_source_capital_cost = Var(
+        initialize=1e6,
+        doc="Capital cost of recarbonation tank - liquid CO2 as CO2 source",
+        domain=NonNegativeReals,
+        units=blk.costing_package.base_currency,
+    )
+
+    blk.lime_feed_system_capital_cost = Var(
```

Comment:
Did costs for this originate from McGiveny and Kawamura? In any case, I think it'd be nice to include the reference maybe in the doc string towards the beginning of the file (or above this var if there are various refs)

Reply:
Reference for cost function was included in the costing block.